cluster.go

```diff
@@ -459,13 +474,73 @@ func (c *Cluster) Do(a Action) error {
 	return c.doInner(a, addr, key, false, doAttempts)
 }
 
+func (c *Cluster) getClusterDownSince() int64 {
+	return atomic.LoadInt64(&c.lastClusterdown)
+}
+
+func (c *Cluster) setClusterDown(down bool) (changed bool) {
+	prevVal := atomic.LoadInt64(&c.lastClusterdown)
+	// avoid CAS in the common case were the cluster is up
+	if !down && prevVal == 0 {
+		return false
+	}
+
+	var newVal int64
+	if down {
+		newVal = time.Now().UnixNano() / 1000 / 1000
+		// check for at least a millisecond difference, to avoid unnecessary stores
+		if newVal-prevVal <= 0 {
+			return false
+		}
+	}
+
+	for !atomic.CompareAndSwapInt64(&c.lastClusterdown, prevVal, newVal) {
+		updatedVal := atomic.LoadInt64(&c.lastClusterdown)
+		// if another goroutine changed the state before we could, we can just return
+		// and avoid more CAS operations since the important thing (the state change)
+		// already happened and the difference in newVal and updatedVal will be small
+		// (when the cluster is down)
+		if prevVal == 0 || updatedVal == 0 || newVal == updatedVal {
```

Comment:
I'm having trouble following this part. Why are we able to assume there wasn't a change if `prevVal` is 0? On its own that doesn't seem like it would have any bearing on whether or not a change had happened.

Reply:
Added a comment that explains this, but I realized that this isn't really necessary since we can just use a swap which is much simpler and easier to get right and has basically the same performance. But yeah the commit with the comment is still there.